rfd/0144-client-tools-updates.md

```diff
@@ -0,0 +1,309 @@
+---
+authors: Russell Jones (rjones@goteleport.com) and Bernard Kim (bernard@goteleport.com)
+state: draft
+---
+
+# RFD 0144 - Client Tools Updates
+
+## Required Approvers
+
+* Engineering: @sclevine && @bernardjkim && @r0mant
+* Product: @klizhentas || @xinding33
+* Security: @reedloden
+
+## What/Why
+
+This RFD describes how client tools (like `tsh` and `tctl`) can be kept up to
+date, either using automatic updates or self-managed updates.
+
+Keeping client tools updated helps with security (fixes for known security
+vulnerabilities are pushed to endpoints), bugs (fixes for resolved issues are
+pushed to endpoints), and compatibility (users no longer have to learn and
+understand [Teleport component
+compatibility](https://goteleport.com/docs/upgrading/overview/#component-compatibility)
+rules).
+
+## Details
+
+### Summary
+
+Client tools (like `tsh` and `tctl`) will automatically download and install
+the version of client tools recommended by the Teleport cluster.
+
+The `cluster_maintenance_config` resource will be updated to include a
+`tools_version` field to control the version of client tools advertised by the
+cluster. This resource will not be configurable on Teleport Cloud. It will be
+managed by the Teleport Cloud team.
+
+If cluster administrators want to opt-out of automatic updates and manage
+updates themselves, they will be able to watch the `cluster_maintenance_config`
+resource for changes and push out updates to endpoints manually.
+
+Inspiration drawn from https://go.dev/doc/toolchain.
+
+### Implementation
+
+#### Client tools
+
+##### Automatic updates
+
+The first time a user executes `tsh login` with a version of `tsh` that
+supports automatic updates, the user will be prompted to enroll in automatic
+updates. This choice will be saved in client tools configuration.
```

Comment:
Does it make sense to allow individual users to opt-out? If you're a Teleport customer you might want to ensure that all your clients are enrolled into AU, or no client is enrolled into AU.

For the second case (enforce off company-wide), they have `tools_version: off` (except maybe for the cloud?), but for the first case (enforce company-wide), the RFD offers no way to enforce client AU.

Reply:
Hrm, what do you guys think about the following for `cluster_maintenance_config`:

* `tools_version: X.Y.Z`: If an exact version is specified, opt clients into automatic updates and install version `X.Y.Z`. This is how Cloud will operate.
* `tools_version: auto`: The new default for self-hosted clusters. Prompt the user to opt-in to automatic updates.
* `tools_version: off`: Do not enforce client tools updates.

Reply:
It would work for simple use-cases but I think it would not support a Cloud customer wanting to install teleport from a trusted channel, like you described later in the RFD.

I recommend decoupling the version served, and the automatic update policy, as those are two different concepts. For example:

``` yaml
tools_version: x.y.z|auto|off
tools_update_policy: always|never|optin
```

A customer wanting to update clients on their own with an MDM but stream the version from the Teleport CMC events would be able to set `tools_update_policy: never` while we (Cloud) keep advertising the exact version needed to interact with Teleport.

Reply:
Talked with the Product team about this, everyone was onboard with enforcing automatic updates by default.

What do you guys think about this?

```yaml
kind: cluster_maintenance_config
spec:
  autoupdate: on|off
```

```yaml
kind: autoupdate_version
spec:
  tools_version: X.Y.Z|auto
```

This would allow turning off client tools automatic updates at the endpoint in `~/.tsh/config/config.yaml` or cluster via `cluster_maintenance_config`. We have to split it up into two resources because we want to prevent Cloud users from changing `tools_version` but still allow them to turn automatic updates on/off.

What do you guys think?

Reply:
This looks good to me. My only question would be about what `spec.autoupdate` does. Does this applies for tools? agents? both?

MDM customers might want to disable tool auto-update while we still want to enforce agent auto-update by default for any cloud setup.

Do we want to split the CMC field in two:

```yaml
kind: cluster_maintenance_config
spec:
  tools_auto_update: on|off # can be configured by customers, even on cloud
  
  # out of this RFD scope but we might want to leave room for this flag to happen
  agents_auto_update: on|off # can be configured by customers, but not on cloud
```